### ISSUE TITLE
Evaluate Kubernetes 1.28 changes

### ISSUE THREAD

**Issue by lfrancke** [stackabletech/issues#420]
I would like to know if there are any changes in Kubernetes 1.28 that the SDP could potentially make use of (e.g. sidecars), will require changes or will break things on our end.

This should be timeboxed and take up at most 4h of research and reading.
The result of this should be a comment on this issue or follow-up issues listing the high level points of things we could, should and should not do.

Must read:
- https://kubernetes.io/blog/2023/08/15/kubernetes-v1-28-release/

Can read:
- https://kubernetes.io/blog/2023/08/18/retroactive-default-storage-class-ga/
- https://kubernetes.io/blog/2023/08/16/kubernetes-1-28-non-graceful-node-shutdown-ga/
- https://kubernetes.io/blog/2023/08/28/kubernetes-1-28-feature-mixed-version-proxy-alpha/
- https://kubernetes.io/blog/2023/08/24/swap-linux-beta/
- https://kubernetes.io/blog/2023/08/25/native-sidecar-containers/
- https://kubernetes.io/blog/2023/08/23/kubelet-podresources-api-ga/
- https://kubernetes.io/blog/2023/08/21/kubernetes-1-28-jobapi-update/

**Comment by maltesander** [comment]
My summary. I tried to roughly order it descending in terms of importance for us (IMHO). Feedback welcome!

### Cluster admin related

- Control plane (kube-apiserver, kube-scheduler, kube-controller-manager, cloud-controller-manager) and nodes (kubelet and kube-proxy) are now compatible for the last 3 minor (from 2) versions 
  - Updating the cluster once each calendar year will still be within upstream support
  - Less upgrades and node drainage required --> better for long running tasks e.g. huge Trino queries
- Mixed version proxy as **alpha** feature https://kubernetes.io/blog/2023/08/28/kubernetes-1-28-feature-mixed-version-proxy-alpha/
  - hides skew from clients when a local API server cannot serve a certain version and will forward to another API server that can
  - applies during up or downgrades
  - Does not apply to CRDs 
- Node `podresources` API Graduates to GA 
- Enabling swap space on linux now **beta** feature (mostly performance tuning)

### Highlights for developers

- Improvements to CRD validation rules 
  - As **beta** feature, CEL(https://github.com/google/cel-go) can be used to validate custom resources (provided in the CRD)
  - This can be used for non-trival validation (and may not require more complex admission validation webhooks anymore)
  - Validating admission policy (https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/) now **beta** feature.
    - Requires `admissionregistration.k8s.io/v1beta1` API group to be enabled
    - Requires `ValidatingAdmissionPolicy` feature gate in the control plane to be enabled
  - `matchConditions` (https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-matchconditions) with CEL now **beta** feature
  --> This might make the product config in its current state completely obsolete
- Recovery from non-graceful node shutdown now **stable** feature https://kubernetes.io/docs/concepts/architecture/nodes/#non-graceful-node-shutdown 
  - `NodeOutOfServiceVolumeDetach` locked to true on `kube-controller-manager` and cannot be disabled
  - Only affects stateful apps
- API awareness of sidecar containers as **alpha** (https://kubernetes.io/docs/concepts/workloads/pods/init-containers/#api-for-sidecar-containers)
  - New field `restartPolicy` to indicate that an init container is a sidecar container 
  - The kubelet will only wait for the sidecar container to have started instead of waiting for it to finish
  - `restartPolicy` will default to `Always` for init containers. Omitting the field means that you want a true init container that runs to completion before application startup.
  - Once the sidecar has started, the sidecar will be restarted if there is a failure even if the Pods `restartPolicy` is `Never` or `OnFailure` 
  - Sidecars do not block Pod completion: if all regular containers are complete, sidecar containers in that Pod will be terminated. Furthermore, sidecar containers will be restarted (on failure or on normal exit) even during Pod termination.
  --> We use a couple of sidecar containers. We should have a closer look here once it becomes beta
- New pod replacement policy for Jobs as **alpha** feature
  - New field for the Job API that allows you to specify if you want the control plane to make new Pods as soon as the previous Pods begin termination (existing behavior), or only once the existing pods are fully terminated (new, optional behavior).
  - Avoids pod clashes and resource problems
  --> We have resources troubles from time to time, might be interesting to have a look.
- Job retry backoff limit, per index as **alpha** feature
  - This extends the Job API to support indexed jobs where the backoff limit is per index, and the Job can continue execution despite some of its indexes failing
  - Currently, the indexes of an indexed job share a single backoff limit. When the job reaches this shared backoff limit, the job controller marks the entire job as failed, and the resources are cleaned up, including indexes that have yet to run to completion which did not cover embarrassingly parallel use cases, where each index is fully independent of other indexes
  --> This could be interesting for Spark or the first phase of map-reduce tasks
- Automatic, retroactive assignment of a default `StorageClass` graduates to **stable**
  - Kubernetes automatically sets a `storageClassName` for a `PersistentVolumeClaim` (PVC) if you don't provide a value
  - The control plane also sets a `StorageClass` for any existing PVC that doesn't have a storageClassName defined  

### Plugin developers  

- Support for CDI injection into containers as **alpha** (no real idea what this actually is, just for completeness).

**Comment by lfrancke** [comment]
Thank you. What I'm missing is how any of those apply to **us**

Which of these features do we need to teach our operators? What should we document ("if you use X be aware that...") etc.

**Comment by maltesander** [comment]
I tried making some proposals/hints (look for the "-->" after the points) regarding the developer highlights. I can go more into detail though tomorrow.
But i would say everything from the developer highlights will apply to us.

**Comment by sbernauer** [comment]
TLDR: LGTM
I have read through the changes @maltesander mentioned and I'm happy with the list.
IMHO `Validating admission policy` is the only one we should invest time into. But it's beta only, and I don't want users to have to enable a feature gate. So for me personally nothing in 1.28 we should care about.

**Comment by dervoeti** [comment]
One more notice regarding a concrete use case for us, because I just stumbled upon it:
Vector would be a perfect fit for a native sidecar container. They even explicitly list this use case in the blog post (https://kubernetes.io/blog/2023/08/25/native-sidecar-containers/#when-to-use-sidecar-containers, "Log collection containers"). It would make sure that the Vector sidecar is completely loaded and ready before any workload container starts. Also, it enables Vector to collect logs of init containers, which is not possible at the moment.
Unfortunately, sidecar containers are also behind a feature gate in 1.28, it will likely take some time until it's really usable for us.
